doc: improve the description of the `decide equality' tactic (#16461)
coq/coq: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

.. tacn:: decide equality

>    Solves a goal of the form :g:`forall x y : R, {x = y} + {~ x = y}` or
>    :g:`forall x y : R, (x = y) \/ (~ x = y)`, where :g:`R` is an inductive type
>    such that its constructors do not take proofs or functions as arguments, nor
>    objects in dependent types. It solves goals of the form
>    :g:`{x = y} + {~ x = y}` or :g:`(x = y) \/ (~ x = y)` as well.

Reviewer: I've hard wrapped it for you:
```suggestion
   Solves a goal of the form :g:`forall x y : R, {x = y} + {~ x = y}` or
   :g:`forall x y : R, (x = y) \/ (~ x = y)`, where :g:`R` is an
   inductive type such that its constructors do not take proofs
   or functions as arguments, nor objects in dependent types. It
   solves goals of the form
   :g:`{x = y} + {~ x = y}` or :g:`(x = y) \/ (~ x = y)` as well.
```

Author: I'm sorry, did I miss the maximum line size somewhere in the contributor documentation?  I've already wrapped it using the emacs macro (M-q).

Reviewer: It's not particularly important for the documentation, however it does make reviewing on Github easier. I also wrapped using emacs, but mine seems to have done it tighter (I'm guessing this is 80).

Author: OK.

.. tacn:: decide equality

   Solves a goal of the form :g:`forall x y : R, {x = y} + {~ x = y}` or
   :g:`forall x y : R, (x = y) \/ (~ x = y)`, where :g:`R` is an

>    inductive type such that its constructors do not take proofs
>    or functions as arguments, nor objects in dependent types. It

Reviewer 2: ```suggestion
   inductive type whose constructors do not take proofs
   or functions as arguments, nor objects in dependent types. It
```

Optiona: Is there better wording for "proofs or functions as arguments, nor objects in dependent types"?  It's a bit opaque to me.  I don't recall other references to proofs as arguments in the doc--proofs are just terms.  It's also better to state what is accepted first before saying what's not accepted.  "nor objects in dependent types" - in arguments or not in arguments?  (Note I've not used and don't fully understand the tactic.  Or dependent types.)  @Alizter perhaps you can suggest better wording?

Reviewer 2: @Alizter FWIW, probably better if you had made your last comment [below](https://github.com/coq/coq/pull/16461#issuecomment-1244092632) as a reply to the review comment, closer to what it responds to.

`(B a) -> I` is functions, makes sense
`A : Prop |- A -> I` looks like a proof but I don't see `|-` in the grammar for terms nor do I see it defined as a notation.
Then `(a -> b) -> I` must be "objects in dependent types", which I don't understand.  What is an "object"?  That's not common terminology.  Also, how do you know if it's dependent or non-dependent?  Perhaps that depends on what a and b are.

@Zimmi48 Perhaps you would suggest some wording?

Or we can just leave it as it is for now.

Reviewer: @jfehrle Sorry `|-` is just short hand for the context. So I meant something like `forall (A : Prop), A -> I` or `A -> I` where `A : Prop` in the context.

Reviewer: `B a` is "objects in dependent types". (`B` is dependent on the value of `a`)
`(a -> b)` is functions. (the type of functions `a -> b`).

   solves goals of the form

>    :g:`{x = y} + {~ x = y}` or :g:`(x = y) \/ (~ x = y)` as well.

Reviewer 2: Just curious: what will this tactic do with a goals in the form `{~ x = y} + {x = y}`?  If it doesn't handle it, maybe add a sentence on how transform it to the form it can handle if that's simple but not totally obvious.

>    :g:`{x = y} + {~ x = y}` or :g:`(x = y) \/ (~ x = y)` as well.

Author: I just tried, and it does handle goals in this form.

>    :g:`{x = y} + {~ x = y}` or :g:`(x = y) \/ (~ x = y)` as well.

Author: I almost missed your suggestion above…  I will apply it.
